### PR TITLE
fix programmatic focus/blur events, typing into currently focused

### LINKED ISSUE #1176: Blur event is not fired reliably in some situations
# bug

Mac, Cypress 1.4.1, Electron and Chrome 63

I have investigated source of failures in Todomvc app some more and found interesting thing. Take a look at https://github.com/cypress-io/cypress-test-tiny/tree/blur-test (I put entire Ampersand TodoMVC there in the branch `blur-test`, you can just open ampersand/index.html to see and test).

The test adds 3 todos, edits second, then checks if the local storage has been updated with new title text.

```js
it('edits todo', () => {
  createDefaultTodos()
  // let everything settle
  cy.wait(1000)

  visibleTodos()
    .eq(1)
    .find('label')
    .dblclick()

  // clear out the inputs current value
  // and type a new value
  visibleTodos()
    .eq(1)
    .find('.edit')
    .should('have.value', 'bar')
    .clear()
    .type('buy some sausages{enter}')
  cy.wait(1000)

  // confirm that new value is in localStorage
  cy
    .window()
    .its('localStorage')
    .its('todos-ampersand')
    .then(JSON.parse)
    .should('have.length', 3)
    .its('1')
    .its('title')
    .should('equal', 'buy some sausages')
})
```

When running in Electron or Chrome browser with Devtools closed, the test passes. I can open DevTools AFTER the test and see the "enter" + "blur" events triggered and propagating into the app.

screen shot 2018-01-09 at 3 10 24 pm

screen shot 2018-01-09 at 3 10 35 pm

If I open DevTools but keep the FOCUS on the app iframe or on the command log and reload the tests, they still work. But if I click on the devtools, for example into the console tab and reload the tests, the "blur" event is not fired anymore and the test fails

screen shot 2018-01-09 at 3 11 22 pm

Probably related to blur event is the run mode. It just fails 100%

screen shot 2018-01-09 at 3 12 15 pm

If I could see console.log messages during this run (https://github.com/cypress-io/cypress/issues/448) it could shed some light

### LINKED ISSUE #1486: react-select v2 dropdown issues
- Operating System: Local: Mac OS High Sierra 10.13.3, CI: Travis (Open source -- https://travis-ci.org)
- Cypress Version: 2.1.0
- Browser Version: Local: Chrome 65.0.3325.181 

### Is this a Feature or Bug?
I think it's a bug, but it may be my mistake.

### Current behavior:
The test that fails is clicking a dropdown and then trying to select an element in that dropdown. It is using a third party open source dropdown component, [react select](https://github.com/JedWatson/react-select). Running this test locally, it passes over and over. I have not witnessed the test failing locally. Running the test through CI, it has failed every time. The video recordings on my Cypress dashboard suggest that the dropdown is never opening as a result of clicking on the element. I have two other tests in my test suite which both pass reliably on CI.

### Desired behavior:
I would like the test pass on CI just as it does locally.

### How to reproduce:
I have an open source repository where I am trying to add cypress. Here is the pull request: 
* https://github.com/cjlitch/screenshot-test-app/pull/15.

Here is the latest commit where you should be able to clone and reproduce the issue:
* https://github.com/cjlitch/screenshot-test-app/tree/20fdd95168b3febd76863251a697b544bf299e5c. 

Here is the relevant Cypress test: 
* https://github.com/cjlitch/screenshot-test-app/blob/20fdd95168b3febd76863251a697b544bf299e5c/cypress/integration/e2e-cypress.js#L17

#### Failing on CI
image

#### Passing Locally
image

### LINKED ISSUE #3001: cy.type shouldn't perform actionability if subject is the currently focused element
If the subject to cy.type is focused, we shouldn't perform actionability because the user is already able to start sending keys to the element without needing to click/tab into it.

This allows apps to call .focus() on an otherwise covered/animating element and allow cypress to still type into it, just as a user would

### LINKED ISSUE #4039: Visibility errors when window is hidden during headed mode test execution
### Current behavior:

Cypress tests fail when the browser is not focused in "run headed" mode. 

This kind of problem is always related with components with animations and overlays like modals, popups or sticky headers. (this example shows the problem with sweet alert library). 

### Desired behavior:

The tests should be deterministic no matter the execution mode and the browser focus state. 

### Steps to reproduce: (app code and test code)

Repository to reproduce the problem (check readme to follow instructions)
https://github.com/agualis/flaky-swal-cypress

Video showing the error:  
https://github.com/agualis/flaky-swal-cypress/blob/master/flaky-cypress-test.mov

### Versions

Cypress:    3.2.0                                                                              │
Browser:    Chrome 73 
OS: macOS Mojave

### PR DESCRIPTION
**Pull request by Bkucera**
- react-select v2 (programmatic focus events while browser out of focus)
fix #1486 

- programmatic blur events while browser out of focus
fix #1176 

- typing into currently focused
fix #3001 

- Visibility errors when window is hidden during headed mode test execution
  fix #4039 

----
#### tasks
- [x] intercept calls to `.blur` + test
- [x] add comments on tests referencing issues
- [x] add test for focusing on already active element
- [x] make `document.hasFocus()` always return true
- [x] follow all the rules on whether element should fire focus events (is focusable, not activeElement)
- [x] move test from focus_spec to type_spec
- [x] add more specific element tests to focus_blur_spec
- [x] add tests for `selectionchange` event to focus_blur_spec
- [x] add intercept for SVGElement focus/blur

#### review tasks
- [x] fix package.json dep version

### CODE REVIEW

**Comment by brian-mann** [comment]
We need to work on this PR a bit and clean it up before merging. Also we need some issues that this PR fixes. Wasn't there one or more issues you went through on the existing `cy.click()` issues that this fixed?

File: packages/driver/src/cypress/cy.coffee [hunk 1]
```diff
@@ -727,6 +731,7 @@ create = (specWindow, Cypress, Cookies, state, config, log) ->
         window: s.window
         document: s.document
         $autIframe: s.$autIframe
+        actAsIfWindowHasFocus: s.actAsIfWindowHasFocus
```
Comment (brian-mann) [inline packages/driver/src/cypress/cy.coffee]: this appears unnecessary

**Comment by brian-mann** [comment]
Also it appears this PR is fixing two things - 1 being forcing the focus/blur events, the other finding the focusable element in the situation such as `[contenteditable]`

We'll need to reference multiple issues here.

File: packages/driver/src/cypress/cy.coffee (hunk 1 above)
Comment (Bkucera) [reply packages/driver/src/cypress/cy.coffee]: without it, `state('actAsIfWindowHasFocus')` was always false for me

**Comment by Bkucera** [comment]
updated the issues, should probably add a test for call to `.blur` however

**Comment by Bkucera** [comment]
also, we shouldn't be using `callNativeMethod` unless we're trying to simulate a default action. So if AUT calls .focus, we should call the `focus` method in the AUT, *not* the one on window.top

**Comment by PhilipGrey** [comment]
Looking forward to this.

File: packages/driver/package.json [hunk 2]
```diff
@@ -31,6 +31,7 @@
     "bytes": "3.1.0",
     "chai": "3.5.0",
     "chai-as-promised": "6.0.0",
+    "chai-subset": "^1.6.0",
```
Comment (jennifer-shehane) [inline packages/driver/package.json]: Can you set this to exact version downloaded plz

File: packages/driver/src/cy/commands/actions/type.coffee [hunk 3]
```diff
@@ -328,16 +328,29 @@ module.exports = (Commands, Cypress, cy, state, config) ->
...
+        $focused = cy.getFocused()
+        $focused = $focused && $focused[0]
+
+        if $elements.isFocusable(options.$el)
+          elToCheckCurrentlyFocused = options.$el[0]
+        else if $elements.isContentEditable(options.$el[0])
+          elToCheckCurrentlyFocused = $selection.getHostContenteditable(options.$el[0])
+
+        if elToCheckCurrentlyFocused && elToCheckCurrentlyFocused is $focused
+          return type()
```
Comment (Bkucera) [inline packages/driver/src/cy/commands/actions/type.coffee]: add TODO: not scrolling here, but revist when scroll algorithm changes

**Pull request merged**